Implement card carousel

Jay-flow opened the pull request:
It's still in progress, so I'll share the interim process.

![image](https://user-images.githubusercontent.com/29420674/127887149-13aafc21-e1eb-44e8-8910-931c6abf9aa2.png)

There are parts where Props and CSS have not been cleaned up yet.

---
And I'm having a problem, so if you know how to solve it, please share it with me.
My problem is to use tyscript to define the type of object. I want to define the type and export it and use it for other components, but the error below appears.

![Screenshot from 2021-08-03 00-47-11](https://user-images.githubusercontent.com/29420674/127888851-92c2705a-04d4-4c2b-b80b-a3a0c86a20be.png)

Do you happen to know how to solve this?
I've referred to the URL below for how to export.
[https://stackoverflow.com/questions/64064506/export-typescript-type-in-svelte-file](https://stackoverflow.com/questions/64064506/export-typescript-type-in-svelte-file)

DevYuns commented:
@Jay-flow It's because typescript limitations with svelte. 
```
when transpiling, it's not possible to import types or interfaces into your svelte component without using the new TS 3.8 type import modifier: import type { SomeInterface } from './MyModule' otherwise bundlers will complain that the name is not exported by MyModule.
```

you can check details [here](https://github.com/sveltejs/svelte-preprocess/blob/main/docs/preprocessing.md#typescript---limitations).

But when you use `import type {Type}` and `import {Component}` in same place, it should throw a `duplicate error`. So what about make a `types.svelte` inside `carousel` folder? You can split types and component. It works well 😄 .

DevYuns commented:
> It's still in progress, so I'll share the interim process.
> 
> ![image](https://user-images.githubusercontent.com/29420674/127887149-13aafc21-e1eb-44e8-8910-931c6abf9aa2.png)
> 
> There are parts where Props and CSS have not been cleaned up yet.
> 
> And I'm having a problem, so if you know how to solve it, please share it with me.
> My problem is to use tyscript to define the type of object. I want to define the type and export it and use it for other components, but the error below appears.
> 
> ![Screenshot from 2021-08-03 00-47-11](https://user-images.githubusercontent.com/29420674/127888851-92c2705a-04d4-4c2b-b80b-a3a0c86a20be.png)
> 
> Do you happen to know how to solve this?
> I've referred to the URL below for how to export.
> https://stackoverflow.com/questions/64064506/export-typescript-type-in-svelte-file

@Jay-flow It's because typescript limitations with svelte. 
```
when transpiling, it's not possible to import types or interfaces into your svelte component without using the new TS 3.8 type import modifier: import type { SomeInterface } from './MyModule' otherwise bundlers will complain that the name is not exported by MyModule.
```

you can check details [here](https://github.com/sveltejs/svelte-preprocess/blob/main/docs/preprocessing.md#typescript---limitations).

But when you use `import type {Type}` and `import {Component}` in same place, it should throw a `duplicate error`. So what about make a `types.svelte` inside `carousel` folder? You can split types and component. It works well 😄 .

hyochan commented:
@Jay-flow Try to add styles generated in figma.
<img width="820" alt="Screen Shot 2021-08-03 at 12 33 05 PM" src="https://user-images.githubusercontent.com/27461460/127953441-1e512cf9-af7f-4103-b6d5-8695a5f584ba.png">

It is not hard to style it.

Jay-flow commented:
@hyochan
I have something to ask about your help last time.

[[CardCarousel] specification](https://github.com/wecount-dev/wecount/issues/16#issuecomment-890657025)

In this issue, you told me to inject the color array from the external component.
In this case, the component needs to give a prop that selects a color among the candidate colors.

What do you think about injecting color code directly into a prop?
I think this is the way to decide the color more neatly.
For example, the code below is the same.

```
<script lang="ts">
  import Carousel from '../uis/Carousel/Carousel.svelte';

  let string = `This is <strong>Temp!</strong>`;

  const item = {
    user: {
      name: 'Jay-flow',
      imageUrl: 'https://source.unsplash.com/random/50x50',
      role: '최고 관리자',
    },
    community: {name: 'dooboolab', description: 'PREMIUM ACCOUNT'},
    isPublic: true,
  };

  const items = [
    {...item, selectedColor: '#28DB98'},
    {...item, selectedColor: '#72E6FF'},
    {...item, selectedColor: '#3A74E7'},
    {...item, selectedColor: '#834FF3'},
    {...item, selectedColor: '#2A2A2C'},
  ];
</script>

<div>
  <p>{@html string}</p>

  <div>
    <Carousel items={items} />
  </div>
</div>

```

Jay-flow commented:
![Screenshot from 2021-08-03 20-36-03](https://user-images.githubusercontent.com/29420674/128008913-7bd0a65f-c688-4c08-bfae-7833d46796ea.png)

And we shared our opinion last time that up to 3 dots are displayed at the maximum. If I have more than 3 cards, should I make it looped? For example, for the fourth card, the activation of the point is the first again.

hyochan commented:
> @hyochan
> I have something to ask about your help last time.
> 
> [[CardCarousel] specification](https://github.com/wecount-dev/wecount/issues/16#issuecomment-890657025)
> 
> In this issue, you told me to inject the color array from the external component.
> In this case, the component needs to give a prop that selects a color among the candidate colors.
> 
> What do you think about injecting color code directly into a prop?
> I think this is the way to decide the color more neatly.
> For example
> 
> ```
> <script lang="ts">
>   import Carousel from '../uis/Carousel/Carousel.svelte';
> 
>   const item = {
>     user: {
>       name: 'Jay-flow',
>       imageUrl: 'https://source.unsplash.com/random/50x50',
>       role: '최고 관리자',
>     },
>     community: {name: 'dooboolab', description: 'PREMIUM ACCOUNT'},
>     isPublic: true,
>   };
> 
>   const items = [
>     {...item, selectedColor: '#28DB98'},
>     {...item, selectedColor: '#72E6FF'},
>     {...item, selectedColor: '#3A74E7'},
>     {...item, selectedColor: '#834FF3'},
>     {...item, selectedColor: '#2A2A2C'},
>   ];
> </script>
> 
> <div>
>   <Carousel items={items} />
> </div>
> ```

I don’t think we need the same data for all cards they should all be the same.

```

const item = {
  user: {
    name: 'Jay-flow',
    imageUrl: 'https://source.unsplash.com/random/50x50',
    role: '최고 관리자',
  },
  community: {name: 'dooboolab', description: 'PREMIUM ACCOUNT'},
  isPublic: true,
};

```

The only variants in the carousel are the colors which I think can be supported as an array.

Jay-flow commented:
@hyochan 
It was a question that occurred because I didn't understand the plan intention well.
All the other data are the same, but only the color can be selected by the user, right? 
Then I think it should be implemented as you said.

Jay-flow commented:
@hyochan 
It was a question that occurred because I didn't understand the plan intention well.
All the other data are the same, but only the color can be selected by the user, right? 
Then I think it should be implemented as you said.

```
<script lang="ts">
  import Carousel from '../uis/Carousel/Carousel.svelte';

  const item = {
    user: {
      name: 'Jay-flow',
      imageUrl: 'https://source.unsplash.com/random/50x50',
      role: '최고 관리자',
    },
    community: {name: 'dooboolab', description: 'PREMIUM ACCOUNT'},
    isPublic: true,
    colors: ['#28DB98', '#72E6FF', '#3A74E7', '#834FF3', '#2A2A2C'],
  };

  function getColor(selectedColor: string) {
    console.log(selectedColor);
  }
</script>

<div style="padding: 30px; background-color:white;">
  <p>{@html string}</p>

  <div>
    <Carousel item={item} onChange={getColor} />
  </div>
</div>
```

![ezgif com-gif-maker](https://user-images.githubusercontent.com/29420674/128047822-57154101-9c20-45f6-97eb-13cc938f00ab.gif)

Jay-flow commented:
@hyochan @DevYuns @KoreanThinker 
Is there any more function you want me to add?

hyochan reviewed (changes requested):
Overall~!! Very good starting point!
I think we need extra style for the `container` so that we can set the margin from parent!

DevYuns commented on src/components/uis/Carousel/Carousel.svelte:
```diff
@@ -0,0 +1,137 @@
+  export let onChange: (selectedColor: string) => void;
+  const colors = item.colors as string[];
+
+  type PositionType = {
+    [key: string]: number;
+  };
+
+  let position: PositionType = {
+    left: -1,
+    center: 0,
+    right: 1,
+  };
```
@Jay-flow  How do you feel about using `Enum` instead?

Jay-flow replied:
> @Jay-flow How do you feel about using `Enum` instead?

I used Enum and deleted it. I don't know why, but Svelt compilation keeps warning me when exporting Enum. Do you happen to know how to solve it?

hyochan replied:
@Jay-flow Try to add union types instead. Enum is not a good solution.

For example, `'left' | 'center' | 'right'`

hyochan replied:
Good review tough

DevYuns replied:
Opps! it is related svelte compiler. `eslint` will complain about `unused-export-let` since `enum` will be compiled into `var`. If you use `enum` in component file, there is no error.

DevYuns replied:
@hyochan I saw your review a bit late 😅 . Can I ask you the reason why `union` is better solution in this case?

Jay-flow replied:
@DevYuns I think it's because Enum is more meaningful to use as a separator. That is, the purpose is different.

Jay-flow commented:
If you inject style outside of the carousel component, what do you think about specifying it as below

```
<style lang="postcss">
  :global([style='carousel']) {
    margin-top: 100px;
  }
</style>

<script lang="ts">
  import Carousel from '../uis/Carousel/Carousel.svelte';

  const item = {
    user: {
      name: 'Jay-flow',
      imageUrl: 'https://source.unsplash.com/random/50x50',
      role: '최고 관리자',
    },
    community: {name: 'dooboolab', description: 'PREMIUM ACCOUNT'},
    isPublic: true,
    colors: ['#28DB98', '#72E6FF', '#3A74E7', '#834FF3', '#2A2A2C'],
  };

  function getColor(selectedColor: string) {
    /*eslint no-console: */
    console.log(selectedColor);
  }
</script>

<div>
  <Carousel style="carousel" item={item} onChange={getColor} />
</div>

```
```
<!-- This is Carousel component -->
<div class="carousel" style={style}>
  ...
</div>
```

I referred to [this](https://svelte.dev/repl/c6b1e24a6f0844de8b032b0c23e928ee?version=3.41.0) article.

DevYuns commented:
> @hyochan @KoreanThinker @DevYuns
> 
> If you inject style outside of the carousel component, what do you think about specifying it as below
> 
> ```
> <style lang="postcss">
>   :global([style='carousel']) {
>     margin-top: 100px;
>   }
> </style>
> 
> <script lang="ts">
>   import Carousel from '../uis/Carousel/Carousel.svelte';
> 
>   const item = {
>     user: {
>       name: 'Jay-flow',
>       imageUrl: 'https://source.unsplash.com/random/50x50',
>       role: '최고 관리자',
>     },
>     community: {name: 'dooboolab', description: 'PREMIUM ACCOUNT'},
>     isPublic: true,
>     colors: ['#28DB98', '#72E6FF', '#3A74E7', '#834FF3', '#2A2A2C'],
>   };
> 
>   function getColor(selectedColor: string) {
>     /*eslint no-console: */
>     console.log(selectedColor);
>   }
> </script>
> 
> <div>
>   <Carousel style="carousel" item={item} onChange={getColor} />
> </div>
> ```
> 
> ```
> <!-- This is Carousel component -->
> <div class="carousel" style={style}>
>   ...
> </div>
> ```
> 
> I referred to [this](https://svelte.dev/repl/c6b1e24a6f0844de8b032b0c23e928ee?version=3.41.0) article.

It looks like `CSS-In-JS` 😆. I just wonder what benefit can we have.

KoreanThinker commented:
@Jay-flow 

I used like that 
```svelte
<div>
  <Carousel style="margin-top: 100px;" item={item} onChange={getColor} />
</div>
```
```svelte
<!-- This is Carousel component -->

<script>
  export let style =''
</script>

<div class="carousel" style={style}>
  ...
</div>
```

https://github.com/wecount-dev/wecount/blob/a3745708704f33428aba291a636ad4c993b769f8/src/components/uis/EditText.svelte#L40-L68

Jay-flow commented:
> @DevYuns
> 
> That's a good.. , But I don't think this component needs complicated style modification. So we just provide basic style modifications.
> 
> like that
> 
> ```svelte
> <div>
>   <Carousel style="margin-top: 100px;" item={item} onChange={getColor} />
> </div>
> ```
> 
> ```svelte
> <!-- This is Carousel component -->
> 
> <script>
>   export let style =''
> </script>
> 
> <div class="carousel" style={style}>
>   ...
> </div>
> ```
> 
> This pattern used in EditText.svelte and other components
> 
> https://github.com/wecount-dev/wecount/blob/a3745708704f33428aba291a636ad4c993b769f8/src/components/uis/EditText.svelte#L40-L68

Okay, I added props that can define the style for Carousel and the style for dot.
```
 <Carousel
      carouselStyle={'margin-top: 150px;'}
      dotStyle={'margin-top: 100px;'}
      item={item}
      onChange={getColor}
    />
```

Jay-flow commented:
I think I made a mistake on the rebase. I'm sorry to tarnish our commitment message.  😭 

I checked that there is no problem with the code.
Is there anything else I need to do about Carousel?

hyochan commented:
> this

This is great information 👍

hyochan commented:
> @hyochan @KoreanThinker @DevYuns
> 
> If you inject style outside of the carousel component, what do you think about specifying it as below
> 
> ```
> <style lang="postcss">
>   :global([style='carousel']) {
>     margin-top: 100px;
>   }
> </style>
> 
> <script lang="ts">
>   import Carousel from '../uis/Carousel/Carousel.svelte';
> 
>   const item = {
>     user: {
>       name: 'Jay-flow',
>       imageUrl: 'https://source.unsplash.com/random/50x50',
>       role: '최고 관리자',
>     },
>     community: {name: 'dooboolab', description: 'PREMIUM ACCOUNT'},
>     isPublic: true,
>     colors: ['#28DB98', '#72E6FF', '#3A74E7', '#834FF3', '#2A2A2C'],
>   };
> 
>   function getColor(selectedColor: string) {
>     /*eslint no-console: */
>     console.log(selectedColor);
>   }
> </script>
> 
> <div>
>   <Carousel style="carousel" item={item} onChange={getColor} />
> </div>
> ```
> 
> ```
> <!-- This is Carousel component -->
> <div class="carousel" style={style}>
>   ...
> </div>
> ```
> 
> I referred to [this](https://svelte.dev/repl/c6b1e24a6f0844de8b032b0c23e928ee?version=3.41.0) article.

This is great information 👍

hyochan commented on src/components/navigations/Main.svelte:
```diff
@@ -3,7 +3,6 @@
     grid-template-columns: 1fr;
     grid-template-rows: 56px 1fr;
   }
-
```
I think you may revert this! Recommended putting `enter` after a paragraph or different contents.

Jay-flow replied:
@hyochan It's the same as you said. 85% expression is more intuitive, so I'll modify it.

hyochan commented on src/components/uis/Carousel/Carousel.svelte:
```diff
@@ -0,0 +1,138 @@
+  .left-card {
+    margin: auto;
+    position: absolute;
+    height: calc(100% - 15%);
```
What is the difference here compared to `85%`?

Jay-flow replied:
@hyochan It's the same as you said. 85% expression is more intuitive, so I'll modify it.

hyochan commented on src/components/uis/Carousel/Card.svelte:
```diff
@@ -0,0 +1,101 @@
+
+  .community-name {
+    color: white;
+    font-size: 1.2rem;
```
We should have our own `Typography` related to this!

Jay-flow replied:
@hyochan I will add props named communityNameStyle so that we can change the style of that part :)

hyochan reviewed (changes requested):
Small reviews added

hyochan commented:
@Jay-flow May you also check how this happens?

<img width="604" alt="Screen Shot 2021-08-05 at 9 49 45 PM" src="https://user-images.githubusercontent.com/27461460/128352684-71edca44-3336-45ed-8943-2fd7719462d2.png">

I don't see any import of `Carousel.svelte` on `Temp` page but it's there.

Jay-flow commented:
> @Jay-flow May you also check how this happens?
> 
> <img alt="Screen Shot 2021-08-05 at 9 49 45 PM" width="604" src="https://user-images.githubusercontent.com/27461460/128352684-71edca44-3336-45ed-8943-2fd7719462d2.png">
> 
> I don't see any import of `Carousel.svelte` on `Temp` page but it's there.

I deleted what was in Temp to merge into Master. I restored it again, so please check it.
I've revised what I said above you as well. :)

hyochan reviewed (approved):
LGTM!

DevYuns reviewed (approved):
Great job!!

Jay-flow commented:
@hyochan @DevYuns @KoreanThinker 
 
There are some updates and changes.
1. Icon file name changed from CamelCase to KebabCase.
2. Changed the file structure of the component associated with the Carousel component.
3. Localization added to the card component.
4. Added a lock icon has been to the card component.
5. Added balance props  to the card component.

![Screenshot from 2021-08-07 19-28-46](https://user-images.githubusercontent.com/29420674/128597430-836c5140-e2d2-429a-b264-da0a12e0c0b2.png)

hyochan commented:
> @hyochan @DevYuns @KoreanThinker
> 
> There are some updates and changes.
> 
> 1. Rename Icon file from Camel case to Kebab case.
> 2. Changed the file structure of the component associated with the Carousel component.
> 3. Added localization  to the card component.
> 4. Added a lock icon has been to the card component.
> 5. Added balance props  to the card component.
> 
> ![Screenshot from 2021-08-07 19-28-46](https://user-images.githubusercontent.com/29420674/128597430-836c5140-e2d2-429a-b264-da0a12e0c0b2.png)

I am worried on no. 5. The balance will be 0 when it is initially created.

Jay-flow commented:
> > @hyochan @DevYuns @KoreanThinker
> 
> > 
> 
> > There are some updates and changes.
> 
> > 
> 
> > 1. Rename Icon file from Camel case to Kebab case.
> 
> > 2. Changed the file structure of the component associated with the Carousel component.
> 
> > 3. Added localization  to the card component.
> 
> > 4. Added a lock icon has been to the card component.
> 
> > 5. Added balance props  to the card component.
> 
> > 
> 
> > ![Screenshot from 2021-08-07 19-28-46](https://user-images.githubusercontent.com/29420674/128597430-836c5140-e2d2-429a-b264-da0a12e0c0b2.png)
> 
> 
> 
> I am worried on no. 5. The balance will be 0 when it is initially created.

Are you saying that the default value of the balance prop should be zero?


I did not show balance in Carousel component, so if I omit it, I designed it invisible in Card component.

Jay-flow commented:
https://user-images.githubusercontent.com/29420674/128630763-df22356d-65d3-498c-a2a5-c2219ebcc524.mp4

**Added Props to set card layout and card style.**
You can check it in the video.

hyochan commented:
>  2021-08-08.20-26-03.mp4 
> **Added Props to set card layout and card style.**
> You can check it in the video.

Awesome!! Could you kindly rebase your work? I'll quickly merge this.

Jay-flow commented:
@hyochan I'm sure I made a mistake.
I accidentally did Merge and then I did Rebase, and I think my repertoire got twisted afterwards. 😭
I don't know why it says changed file even though it's the same as the Upstream master.
This is not the file I worked on.

How can I fix this?
Should I close these Pull request and do everything over again and do Pull request?

I'm so sorry for the confusion.

DevYuns commented:
> @hyochan I'm sure I made a mistake.
> I accidentally did Merge and then I did Rebase, and I think my repertoire got twisted afterwards. 😭
> I don't know why it says changed file even though it's the same as the Upstream master.
> This is not the file I worked on.
> 
> How can I fix this?
> Should I close these Pull request and do everything over again and do Pull request?
> 
> I'm so sorry for the confusion.

It seems your local master or origin master is not aligned with upstream master. I can give you a basic guideline about how to reduce mistake. 

1. First I recommend you to sync upstream master & local master with `git fetch upstream master` -> `git rebase upstream/master`, then pushing to `origin master`. You should check that three master branch(upstream master, origin master, local master) are aligned.
2. Checkout to your working branch, rebase with your local master with `git rebase master`. then resolve conflict.

Jay-flow commented:
Although there is no difference compared to master brunch, File changes display as if they were changed. This is driving me crazy.
Commit messages have also become very messy. I think it would be better to re-create it with a new brunch and send a full request.

It's because of my ignorance. I'm sorry. I don't make a mistake next time. 😭
